### PR TITLE
Fix dynamics unittests involving finite difference function

### PR DESCRIPTION
**Pull request by edantec**
This PR aims at solving issue https://github.com/Simple-Robotics/aligator/issues/127#issue-2131916127 about an error in the finite difference python function used in unittest. It also adds some missing unittest for MultibodyFreeFwdDynamics derivatives.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -34,6 +34,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Export C++ definitions in CMake config file
 - Fix Doxyfile python bindings directory ([#110](https://github.com/Simple-Robotics/aligator/pull/110))
 - Fix for eigenpy 3.3 ([#121](https://github.com/Simple-Robotics/aligator/pull/121))
+- Fix finite difference function in python unittest
```
Comment (jorisv) [inline CHANGELOG.md:37]: You must move this in *Unreleased* section. Also add the PR link.

File: tests/python/test_continous_dynamics.py [hunk 2]
```diff
@@ -87,6 +87,15 @@ def test_multibody_free():
 
         ode.forward(x0, u0, data)
         ode.dForward(x0, u0, data)
+
+        Jx0 = data.Jx.copy()
+        Ju0 = data.Ju.copy()
```
Comment (jorisv) [inline tests/python/test_continous_dynamics.py:92]: Why do you copy it ?

**Review by jorisv — CHANGES_REQUESTED** [review]

**Comment by ManifoldFR** [comment]
I'm working on this PR at the moment

**Comment by ManifoldFR** [comment]
Setting

```python
x0 = space.rand()
x0[:3] = 0.
```
Did increase the residual. I think there's some kind of high-frequency component that 1st-order finite differences just can't capture.

Figure_1

**Comment by ManifoldFR** [comment]
We'll be switching to centered finite differences then. I'll also be scaling $`\epsilon`$ in the finite differences by the norms of x and u.

**Comment by ManifoldFR** [comment]
For further reference, here's what Crocoddyl does:

* selects the finite difference parameter $`\epsilon = \sqrt{2\epsilon^\text{mach}}`$ where $`\epsilon^\text{mach}`$ is the scalar-type machine precision, which should be `2.22045e-16` for doubles according to the C standard (hence $`\epsilon = 2.1\times 10^{-8}`$)
* tests for numerical tolerance $`tol = \epsilon^{1/3}`$ which is about 0.002762 for doubles (comments in the source code refer to [this dead link](http://www.it.uom.gr/teaching/linearalgebra/NumericalRecipiesInC/c5-7.pdf), which I think is [this file](https://www.grad.hr/nastava/gs/prg/NumericalRecipesinC.pdf))

@jorisv would you agree with the state of this PR?
@edantec do you agree with my analysis of Crocoddyl's behaviour?

**Comment by edantec** [comment]
Looks good to me!

**Comment by ManifoldFR** [comment]
> @ManifoldFR We can also look at the 5 points finite difference made in [roboptim](https://github.com/roboptim/roboptim-core/blob/master/include/roboptim/core/decorator/finite-difference-gradient.hxx#L402-L449)
> 
> It's maybe more precise and the code estimate the truncation error and the rounding error.

That must be very accurate, wow

File: CHANGELOG.md (hunk 1 above)
Comment (edantec) [reply CHANGELOG.md]: Done.

**Review by jorisv — APPROVED** [review]

**Pull request merged**